geetools/ImageCollection.py

```diff
@@ -1049,3 +1058,672 @@ def computeDistance(other):
+
+        return ee.Dictionary.fromLists(eeLabels, eeLabels.map(reduce))
+
+    def datesByRegions(
+        self,
+        band: str,
+        regions: ee.FeatureCollection,
+        label: str = "system:index",
+        reducer: str = "mean",
+        scale: int = 10000,
+        dateProperty: str = "system:time_start",
+    ) -> ee.Dictionary:
+        """Reduce the data for each image in the collection by regions for a single band.
+
+        This method is returning a dictionary with all the regions as keys and their reduced value for each date over the specified region for a specific band as value.
+
+        .. code-block::
+
+            {
+                "region1": {"date1": value1, "date2": value2, ...},
+                "region2": {"date1": value1, "date2": value2, ...},
+                ...
+            }
+
+        Parameters:
+            band: The band to reduce.
+            regions: The regions to reduce the data on.
+            label: The property to use as label for each region. Default is "system:index".
+            reducer: The name of the reducer to use. Default is "mean".
+            scale: The scale in meters to use for the reduction. default is 10000m
+            dateProperty: The property to use as date for each image. Default is "system:time_start".
+
+        Returns:
+            A dictionary with the reduced values for each region and each date.
+
+        Examples:
+            .. code-block:: python
+
+            import ee, geetools
+
+            ee.Initialize()
+
+            collection = (
+                ee.ImageCollection("LANDSAT/LC08/C01/T1_TOA")
+                .filterBounds(ee.Geometry.Point(-122.262, 37.8719))
+                .filterDate("2014-01-01", "2014-12-31")
+            )
+
+            regions = ee.FeatureCollection([
+                ee.Feature(ee.Geometry.Point(-122.262, 37.8719).buffer(10000), {"name": "region1"}),
+                ee.Feature(ee.Geometry.Point(-122.262, 37.8719).buffer(20000), {"name": "region2"})
+            ])
+
+            reduced = collection.geetools.datesByRegions("B1", regions, "name", "mean", 10000, "system:time_start")
+            print(reduced.getInfo())
+        """
+        # aggregate all the dates of the image collection into bands of a single image
+        def to_string(date: ee.Date) -> ee.String:
+            return ee.Date(date).format(JS_DATE_FORMAT)
+
+        dateList = self._obj.aggregate_array(dateProperty).map(to_string)
+
+        # reduce the data for each region
+        image = self._obj.select([band]).toBands().rename(dateList)
+        red = getattr(ee.Reducer, reducer)()
+        reduced = image.reduceRegions(regions, red, scale)
+
+        # create a list of dictionaries for each region and aggregate them into a dictionary
+        values = reduced.toList(regions.size()).map(lambda f: ee.Feature(f).toDictionary(dateList))
+        keys = ee.List(regions.aggregate_array(label))
```

Comment:
he number of values is always limited when we arrive at this stage so let me check if this make such a iff in term of computation. If it does I'll code it you way.

Reply:
I did several runs and it seems they perform exactly the same: 

![image](https://github.com/user-attachments/assets/0cba1fca-ff20-4069-8628-79c4ae03351e)

My guess is that iterative filtering might be the way the build the list in the first place. I think what Noel wax talking about is to use `map` instead of `toList` to avoid people using afor loop type of manipulation. In this case in both our method we are force to retrieve each feature individually in the end to run `toDictionary`. 

I'll keep my implementation for the few character save.